lib/link.js

```diff
@@ -100,11 +107,13 @@ Link.prototype.attach = function() {
 
 Link.prototype.detach = function() {
   var self = this;
+  this._timeouts = undefined; // Disable any re-attachment policy.
   var detachPromise = new Promise(function(resolve, reject) {
     var onError = function(err) { reject(err); };
     self.once(Link.ErrorReceived, onError);
     self.once(Link.Detached, function(info) {
       self.removeListener(Link.ErrorReceived, onError);
+      self.session._removeLink(self);
```

Comment:
I struggled with this. Basically a session should only remove the link when it's given up on itself (either because we called detach or because it's run out of retries). The link is the one who knows when it's given up - and that _doesn't_ coincide with a Detach event. So either the session needs to interrogate the link on whether it might still be trying to attach, or the link needs to tell the session it's done. I chose the latter, but I can see why the former makes sense given our separation of concerns - I can do that.